app/services/preserved_object_handler.rb

```diff
@@ -0,0 +1,159 @@
+# creating a PreservedObject and/or updating check timestamps may require interactions
+#  beyond the single PreservedObject model (e.g. PreservationCopy, PreservationPolicy).
+#  This service class encapsulates logic to keep the controller and the model object
+#    code simpler/thinner.
+# NOTE: performing validation here to allow this class to be called directly avoiding http overhead
+#
+# inspired by http://www.thegreatcodeadventure.com/smarter-rails-services-with-active-record-modules/
+class PreservedObjectHandler
+
+  INVALID_ARGUMENTS = 1
+  VERSION_MATCHES = 2
+  ARG_VERSION_GREATER_THAN_DB_OBJECT = 3
+  ARG_VERSION_LESS_THAN_DB_OBJECT = 4
+  UPDATED_DB_OBJECT = 5
+  UPDATED_DB_OBJECT_TIMESTAMP_ONLY = 6
+  CREATED_NEW_OBJECT = 7
+  DB_UPDATE_FAILED = 8
+
+  RESPONSE_CODE_TO_MESSAGES = {
+    INVALID_ARGUMENTS => "encountered validation error(s): %{addl}",
+    VERSION_MATCHES => "incoming version (%{incoming_version}) matches db version",
+    ARG_VERSION_GREATER_THAN_DB_OBJECT => "incoming version (%{incoming_version}) greater than db version",
+    ARG_VERSION_LESS_THAN_DB_OBJECT => "incoming version (%{incoming_version}) less than db version; ERROR!",
+    UPDATED_DB_OBJECT => "db object updated",
+    UPDATED_DB_OBJECT_TIMESTAMP_ONLY => "updated db timestamp only",
+    CREATED_NEW_OBJECT => "added object to db as it did not exist",
+    DB_UPDATE_FAILED => "db update failed: %{addl}"
+  }.freeze
+
+  include ActiveModel::Validations
+
+  # Note: supplying validations here to allow validation before use, e.g. incoming_version in numeric logic
+  validates :druid, presence: true, format: { with: /\A[a-z]{2}[0-9]{3}[a-z]{2}[0-9]{4}\z/ }
+  validates :incoming_version, presence: true, numericality: { only_integer: true, greater_than: 0 }
+  validates :incoming_size, numericality: { only_integer: true, greater_than: 0 }
+
+  attr_reader :druid, :incoming_version, :incoming_size
+
+  def initialize(druid, incoming_version, incoming_size)
+    @druid = druid
+    @incoming_version = version_string_to_int(incoming_version)
+    @incoming_size = string_to_int(incoming_size)
+  end
+
+  def update_or_create
```

Comment:
my understanding of the spec was that we wanted the storage_root to PCC db existence check to do a single call per druid.  This call would generally just update a timestamp ("yes, it's there, and the versions match), but would occasionally (and first time) need to create the record if it wasn't already in the db.

If this logic isn't here, it would be pushed out to the calling program, I think, which then might have to make multiple calls to the PCC when it encounters anything unexpected.

It's possible i got the spec wrong, or that it changed ... but I think this is what was wanted.